relations/src/gr1cs/constraint_system.rs

```diff
@@ -0,0 +1,773 @@
+                } else {
+                    let value = self.eval_lc(idx)?;
+                    self.lc_assignment_cache.borrow_mut().insert(idx, value);
+                    Some(value)
+                }
+            },
+        }
+    }
+
+    /// Evaluate the linear combination `lc` with the assigned values and return
+    /// the result.
+    fn eval_lc(&self, lc: LcIndex) -> Option<F> {
+        let lc: &LinearCombination<F> = self.lc_map.get(&lc)?;
+        let mut acc = F::zero();
+        for (coeff, var) in lc.iter() {
+            acc += *coeff * self.assigned_value(*var)?;
+        }
+        Some(acc)
+    }
+
+    /// If `self` is satisfied, outputs `Ok(true)`.
+    /// If `self` is unsatisfied, outputs `Ok(false)`.
+    /// If `self.is_in_setup_mode()` or if `self == None`, outputs `Err(())`.
+    pub fn is_satisfied(&self) -> crate::utils::Result<bool> {
+        self.which_predicate_is_unsatisfied().map(|s| s.is_none())
+    }
+
+    /// If `self` is satisfied, outputs `Ok(None)`.
+    /// If `self` is unsatisfied, outputs `Some(s,i)`, where `s` is the label of
+    /// the unsatisfied prediacate and  `i` is the index of
+    /// the first unsatisfied constraint in that predicate.
+    /// If `self.is_in_setup_mode()` or `self == None`, outputs `Err(())`.
+    pub fn which_predicate_is_unsatisfied(&self) -> crate::utils::Result<Option<String>> {
+        if self.is_in_setup_mode() {
+            Err(SynthesisError::AssignmentMissing)
+        } else {
+            for (label, predicate) in self.predicate_constraint_systems.iter() {
+                if let Some(unsatisfied_constraint) =
+                    predicate.which_constraint_is_unsatisfied(self)
+                {
+                    let mut trace: String = "".to_string();
+                    #[cfg(feature = "std")]
+                    {
+                        trace = self.predicate_traces[label][unsatisfied_constraint]
+                            .as_ref()
+                            .map_or_else(
+                                || {
+                                    eprintln!(
+                                        "Constraint trace requires enabling `ConstraintLayer`"
+                                    );
+                                    format!("{} - {unsatisfied_constraint}", &label)
+                                },
+                                |t| format!("{t}"),
+                            )
+                            .to_string();
+                    }
+                    #[cfg(not(feature = "std"))]
+                    {
+                        trace = format!("{} - {unsatisfied_constraint}", label.clone());
+                    }
+                    return Ok(Some(trace));
+                }
+            }
+            Ok(None)
+        }
+    }
+
+    /// Finalize the constraint system (either by outlining or inlining,
+    /// if an optimization goal is set).
+    pub fn finalize(&mut self, outline_instance: bool) {
```

Comment:
Change `outline_instance` to a field in `ConstraintSystem`, and remove it from finalize. Default it to `false`.